### PR TITLE
feat: Jsx import source

### PR DESCRIPTION
**Pull request by andreespirela**

### CODE REVIEW

File: crates/base/src/deno_runtime.rs [hunk 1]
```diff
@@ -231,7 +232,15 @@ impl DenoRuntime {
         let is_user_worker = conf.is_user_worker();
 
         // TODO: check for other potential main paths (eg: index.js, index.tsx)
```
Comment (nyannyacha) [inline crates/base/src/deno_runtime.rs:234]: I guess this comment is no longer valid, right?

**Review by nyannyacha — DISMISSED** [review]
LGTM. I’ve tested in local machine and it worked as normally 😁

**Comment by mcandeia** [comment]
Heys guys are we ready to merge it? @nyannyacha @andreespirela

**Comment by mcandeia** [comment]
I don't know If I did something wrong, but I couldn't make it work using this branch. Do you have a working example?

**Review by laktek — APPROVED** [review]

**Comment by mcandeia** [comment]
Tested with a fresh project but it did not work. Then I tried to rewrite the entire source code adding jsx pragma and it worked. So I think there are some missing configuration yet

**Comment by mcandeia** [comment]
image
Easy to reproduce, just try: https://fresh.deno.dev/docs/getting-started/create-a-project

**Comment by nyannyacha** [comment]
@mcandeia 

Note:

```
DEBUG Specifier "https://deno.land/std@0.216.0/dotenv/parse.ts" was not mapped in import map.
DEBUG Specifier "https://deno.land/std@0.216.0/dotenv/stringify.ts" was not mapped in import map.
DEBUG Sending warning alert CloseNotify
DEBUG Sending warning alert CloseNotify
DEBUG Sending warning alert CloseNotify
DEBUG Sending warning alert CloseNotify
runtime has escaped from the event loop unexpectedly: event loop error: TypeError: Invalid Version: supabase-edge-runtime-0.1.0 (compatible with Deno v1.40.3)
    at parse (https://deno.land/std@0.216.0/semver/parse.ts:18:22)
    at ensureMinDenoVersion (https://deno.land/x/fresh@1.6.8/src/dev/mod.ts:9:23)
    at dev (https://deno.land/x/fresh@1.6.8/src/dev/dev_command.ts:9:3)
    at file:///workspaces/edge-runtime/examples/fresh/index.ts:5:7
    at eventLoopTick (ext:core/01_core.js:64:7)
failed to send request to user worker: request has been cancelled by supervisor
user worker failed to respond: request has been cancelled by supervisor
WorkerRequestCancelled: request has been cancelled by supervisor
    at async Promise.allSettled (index 1)
    at async UserWorker.fetch (ext:sb_user_workers/user_workers.js:70:21)
    at async callWorker (file:///workspaces/edge-runtime/examples/main/index.ts:114:14)
    at async #respond (https://deno.land/std@0.131.0/http/server.ts:219:18) {
  name: "WorkerRequestCancelled"
}
cancelled!
DEBUG CPU time used: 50ms
```

[support-jsx-import-source-mcandeia](https://github.com/supabase/edge-runtime/tree/support-jsx-import-source-mcandeia)

**Comment by andreespirela** [comment]
This is interesting 

> TypeError: Invalid Version: supabase-edge-runtime-0.1.0 (compatible with Deno v1.40.3)

**Comment by nyannyacha** [comment]
After modifying `Deno.version`, I was unable to proceed with the error that `Deno.stat` was not found. 😅

```
...
DEBUG Specifier "https://deno.land/std@0.216.0/fs/walk.ts" was not mapped in import map.
DEBUG Specifier "https://deno.land/std@0.216.0/dotenv/mod.ts" was not mapped in import map.
DEBUG Specifier "https://deno.land/std@0.216.0/dotenv/parse.ts" was not mapped in import map.
DEBUG Specifier "https://deno.land/std@0.216.0/dotenv/stringify.ts" was not mapped in import map.
runtime has escaped from the event loop unexpectedly: event loop error: TypeError: Deno.stat is not a function
    at collectDir (https://deno.land/x/fresh@1.6.8/src/dev/mod.ts:22:29)
    at collect (https://deno.land/x/fresh@1.6.8/src/dev/mod.ts:51:5)
    at dev (https://deno.land/x/fresh@1.6.8/src/dev/dev_command.ts:23:29)
    at file:///workspaces/edge-runtime/examples/fresh/index.ts:5:7
    at eventLoopTick (ext:core/01_core.js:64:7)
failed to send request to user worker: request has been cancelled by supervisor
user worker failed to respond: request has been cancelled by supervisor
WorkerRequestCancelled: request has been cancelled by supervisor
    at async Promise.allSettled (index 1)
    at async UserWorker.fetch (ext:sb_user_workers/user_workers.js:70:21)
    at async callWorker (file:///workspaces/edge-runtime/examples/main/index.ts:114:14)
    at async #respond (https://deno.land/std@0.131.0/http/server.ts:219:18) {
  name: "WorkerRequestCancelled"
}
cancelled!
DEBUG CPU time used: 68ms
```

**Comment by nyannyacha** [comment]
```typescript
// fresh.gen.ts
...
const manifest = {
  routes: {
    "./routes/_404.tsx": $_404,
    "./routes/_app.tsx": $_app,
    "./routes/api/joke.ts": $api_joke,
    "./routes/greet/[name].tsx": $greet_name_,
    "./routes/index.tsx": $index,
  },
  islands: {
    "./islands/Counter.tsx": $Counter,
  },
  // baseUrl: import.meta.url
  baseUrl: "file://./mnt/data/examples/fresh/fresh.gen.ts", // <- You will not able to make it a relative path whatever you do
} satisfies Manifest;

export default manifest;

```

```typescript
import {
    dirname,
    fromFileUrl,
} from "https://deno.land/std@0.222.1/path/mod.ts";

console.log(import.meta.url);
console.log(dirname(fromFileUrl("file://./mnt/data/examples/fresh/fresh.gen.ts")));
```

```
[Info] file:///workspaces/edge-runtime/examples/serve/index.ts
[Info] /mnt/data/examples/fresh // 🙄
```

```typescript
// Note that this is an ABSOLUTE path.
console.log(Deno.readTextFileSync("/mnt/data/examples/fresh/deno.json"));
```

```
DEBUG Snapshot already up to date. Skipping pending resolution.
DEBUG Opening cache /home/vscode/.cache/deno/node_analysis_cache_v1...
runtime has escaped from the event loop unexpectedly: event loop error: NotFound: path not found /mnt/data/examples/serve/deno.json: readfile '/mnt/data/examples/serve/deno.json'
    at Object.readTextFileSync (ext:deno_fs/30_fs.js:832:10)
    at file:///workspaces/edge-runtime/examples/serve/index.ts:3:18
failed to send request to user worker: request has been cancelled by supervisor
user worker failed to respond: request has been cancelled by supervisor
Error in worker connection: connection closed before message completed
DEBUG CPU time used: 0ms
WorkerRequestCancelled: request has been cancelled by supervisor
    at async Promise.allSettled (index 1)
    at async UserWorker.fetch (ext:sb_user_workers/user_workers.js:70:21)
    at async callWorker (file:///workspaces/edge-runtime/examples/main/index.ts:114:14)
    at async #respond (https://deno.land/std@0.131.0/http/server.ts:219:18) {
  name: "WorkerRequestCancelled"
}
cancelled!
```

```typescript
// Note that this is a RELATIVE path.
console.log(Deno.readTextFileSync("./mnt/data/examples/fresh/deno.json"));
```

```
...
DEBUG Snapshot already up to date. Skipping pending resolution.
DEBUG Opening cache /home/vscode/.cache/deno/node_analysis_cache_v1...
[Info] {
  "lock": false,
  "tasks": {
    "check": "deno fmt --check && deno lint && deno check **/*.ts && deno check **/*.tsx",
...
```
😅

**Comment by mcandeia** [comment]
> Command
> 
> ```
> EDGE_RUNTIME_PORT=9998 ./target/debug/edge-runtime start -v --static "./examples/**/*.json"
> ```
> 
> ```ts
> // examples/fresh/fresh.gen.ts
> ...
> const manifest = {
>   routes: {
>     "./routes/_404.tsx": $_404,
>     "./routes/_app.tsx": $_app,
>     "./routes/api/joke.ts": $api_joke,
>     "./routes/greet/[name].tsx": $greet_name_,
>     "./routes/index.tsx": $index,
>   },
>   islands: {
>     "./islands/Counter.tsx": $Counter,
>   },
>   // baseUrl: import.meta.url
>   baseUrl: "file://./mnt/data/examples/fresh/fresh.gen.ts", // <- You will not able to make it a relative path whatever you do
> } satisfies Manifest;
> 
> export default manifest;
> ```
> 
> ```ts
> import {
>     dirname,
>     fromFileUrl,
> } from "https://deno.land/std@0.222.1/path/mod.ts";
> 
> console.log(import.meta.url);
> console.log(dirname(fromFileUrl("file://./mnt/data/examples/fresh/fresh.gen.ts")));
> ```
> 
> ```
> [Info] file:///workspaces/edge-runtime/examples/serve/index.ts
> [Info] /mnt/data/examples/fresh // 🙄
> ```
> 
> ```ts
> // Note that this is an ABSOLUTE path.
> console.log(Deno.readTextFileSync("/mnt/data/examples/fresh/deno.json"));
> ```
> 
> ```
> DEBUG Snapshot already up to date. Skipping pending resolution.
> DEBUG Opening cache /home/vscode/.cache/deno/node_analysis_cache_v1...
> runtime has escaped from the event loop unexpectedly: event loop error: NotFound: path not found /mnt/data/examples/serve/deno.json: readfile '/mnt/data/examples/serve/deno.json'
>     at Object.readTextFileSync (ext:deno_fs/30_fs.js:832:10)
>     at file:///workspaces/edge-runtime/examples/serve/index.ts:3:18
> failed to send request to user worker: request has been cancelled by supervisor
> user worker failed to respond: request has been cancelled by supervisor
> Error in worker connection: connection closed before message completed
> DEBUG CPU time used: 0ms
> WorkerRequestCancelled: request has been cancelled by supervisor
>     at async Promise.allSettled (index 1)
>     at async UserWorker.fetch (ext:sb_user_workers/user_workers.js:70:21)
>     at async callWorker (file:///workspaces/edge-runtime/examples/main/index.ts:114:14)
>     at async #respond (https://deno.land/std@0.131.0/http/server.ts:219:18) {
>   name: "WorkerRequestCancelled"
> }
> cancelled!
> ```
> 
> ```ts
> // Note that this is a RELATIVE path.
> console.log(Deno.readTextFileSync("./mnt/data/examples/fresh/deno.json"));
> ```
> 
> ```
> ...
> DEBUG Snapshot already up to date. Skipping pending resolution.
> DEBUG Opening cache /home/vscode/.cache/deno/node_analysis_cache_v1...
> [Info] {
>   "lock": false,
>   "tasks": {
>     "check": "deno fmt --check && deno lint && deno check **/*.ts && deno check **/*.tsx",
> ...
> ```
> 
> 😅

Sorry, I forgot to mention that I did a few modifications to make it work as is. First, for sake of simplicity, I've removed the static VFS and switched to DenoCompileFileSystem, second, I set the import map inline so those imports will be resolved automatically,

@nyannyacha @andreespirela want me to setup a reproducible repo?

**Comment by nyannyacha** [comment]
Hello @mcandeia 😁

> Sorry, I forgot to mention that I did a few modifications to make it work as is. First, for sake of simplicity, I've removed the static VFS and switched to DenoCompileFileSystem, second, I set the import map inline so those imports will be resolved automatically,

Actually, I posted the above not only for you, but also for andres.
Because I wanted to show that static fs mounting from `./mnt/data` might not work in certain scenarios (like as fresh package 😋).

> And I saw that you guys posted earlier today ours, amazing!! So our chance to get the product of the month walked away, but nice to have you guys as the one that will win this month. Good luck, you have my vote!

Thank you!

**Comment by nyannyacha** [comment]
> want me to setup a reproducible repo?

Yes, That will help 😁
If you create a reproducible repository with changes to this PR, I'd be happy to review it.

**Comment by mcandeia** [comment]
> > want me to setup a reproducible repo?
> 
> Yes, That will help 😁 If you create a reproducible repository with changes to this PR, I'd be happy to review it.

I can certainly do that but it's not working e2e, [somehow its failing when trying to compile .tsx using the specified JsxImportSource](https://github.com/supabase/edge-runtime/pull/301#issuecomment-2053860238), I have no idea why, needs some deeper investigation. But let me prepare a reproducible repo tomorrow and I'll send it back here

**Comment by mcandeia** [comment]
Guys @nyannyacha and @andreespirela, check out this PR:

https://github.com/deco-cx/edge-runtime/pull/6 try out yourself, I did the modifications mentioned before and added a `examples/fresh-project` which you can use to test

**Comment by mcandeia** [comment]
> This is interesting
> 
> > TypeError: Invalid Version: supabase-edge-runtime-0.1.0 (compatible with Deno v1.40.3)

btw, I think this is not a semver compliant version, that should be the way

**Comment by nyannyacha** [comment]
@mcandeia 

Yeah, I didn't realise that property was being parsed as semver by other packages. 😓
I'll have to revisit this.

**Comment by nyannyacha** [comment]
(FYI, I am at GMT+9 and start my activity around this time every day. I will now go through your reproducible repository 😋)

**Comment by nyannyacha** [comment]
@mcandeia 

I looked into the issue in your fork and I think I found what happened.

PR-6/examples/fresh-project/index.tsx (I suffixed an `x` after it to confirm my theory.)
스크린샷 2024-04-18 오전 9 11 01

PR-6/examples/fresh-project/routes/index.tsx
스크린샷 2024-04-18 오전 9 01 53

I don't know what happened, but the emit option specified in the entrypoint (which should have been specified as _jsx, not React) was transpiled without being passed to the subfile.

cc @andreespirela

**Comment by nyannyacha** [comment]
Update:

I've found the problematic line.

https://github.com/supabase/edge-runtime/blob/7014052d91605bf9b0f96dabd46ccfd42fec6a7c/crates/sb_graph/graph_util.rs#L333-L334

```diff
--- a/crates/sb_graph/graph_util.rs
+++ b/crates/sb_graph/graph_util.rs
@@ -330,7 +330,7 @@ pub async fn create_eszip_from_graph_raw(
     let parser_arc = emitter.clone().parsed_source_cache().unwrap();
     let parser = parser_arc.as_capturing_parser();
 
-    eszip::EszipV2::from_graph(graph, &parser, Default::default())
+    eszip::EszipV2::from_graph(graph, &parser, emitter.emit_options())
 }
 
 pub async fn create_graph(

```

I've confirmed that this patch is rendering HTML static content correctly on @mcandeia's PR-6 without file system sandboxing and static fs applied.

Output:
```
<!DOCTYPE html>
<html>

<head>
    <title>fresh-project</title>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" href="/styles.css" />
    <link rel="modulepreload" href="/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/deserializer.js" />
    <link rel="modulepreload" href="/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/signals.js" />
    <link rel="modulepreload" href="/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/main.js" />
    <link rel="modulepreload" href="/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/island-counter.js" />
</head>

<body>
    <div class="px-4 py-8 mx-auto bg-[#86efac]">
        <div class="max-w-screen-md mx-auto flex flex-col items-center justify-center"><img class="my-6" src="/logo.svg?__frsh_c=d9d51bae574d9c3d4563416258222b74e69df4ee" width="128" height="128" alt="the Fresh logo: a sliced lemon dripping with juice"/>
            <h1 class="text-4xl font-bold">Welcome to Fresh</h1>
            <p class="my-4">Try updating this message in the<code class="mx-2">./routes/index.tsx</code> file, and
                refresh.</p>
            <!--frsh-counter_default:0:-->
            <div class="flex gap-8 py-6">
                <button disabled class="px-2 py-1 border-gray-500 border-2 rounded bg-white hover:bg-gray-200 transition-colors">-1</button>
                <p class="text-3xl tabular-nums">3</p>
                <button disabled class="px-2 py-1 border-gray-500 border-2 rounded bg-white hover:bg-gray-200 transition-colors">+1</button>
            </div>
            <!--/frsh-counter_default:0:-->
        </div>
    </div>
    <script id="__FRSH_STATE_4bac2f0c-8d08-4e0a-885a-c51491ebafcc" type="application/json"
        nonce="341dac263ecd40cd91b676254f2f435c">
        {"v":[[{"count":{"_f":"s","v":3}}],[]]}
    </script>
    <script type="module" nonce="341dac263ecd40cd91b676254f2f435c">
        import { deserialize } from "/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/deserializer.js";import { signal } from "/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/signals.js";const ST = document.getElementById("__FRSH_STATE_4bac2f0c-8d08-4e0a-885a-c51491ebafcc").textContent;const STATE = deserialize(ST, signal);import { revive } from "/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/main.js";import counter_default from "/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/island-counter.js";const propsArr = typeof STATE !== "undefined" ? STATE[0] : [];revive({counter_default:counter_default,}, propsArr);
    </script>
</body>

</html>
```

**Comment by mcandeia** [comment]
> Update:
> 
> I've found the problematic line.
> 
> https://github.com/supabase/edge-runtime/blob/7014052d91605bf9b0f96dabd46ccfd42fec6a7c/crates/sb_graph/graph_util.rs#L333-L334
> 
> ```diff
> --- a/crates/sb_graph/graph_util.rs
> +++ b/crates/sb_graph/graph_util.rs
> @@ -330,7 +330,7 @@ pub async fn create_eszip_from_graph_raw(
>      let parser_arc = emitter.clone().parsed_source_cache().unwrap();
>      let parser = parser_arc.as_capturing_parser();
>  
> -    eszip::EszipV2::from_graph(graph, &parser, Default::default())
> +    eszip::EszipV2::from_graph(graph, &parser, emitter.emit_options())
>  }
>  
>  pub async fn create_graph(
> ```
> 
> I've confirmed that this patch is rendering static HTML content correctly on @mcandeia's PR-6 without file system sandboxing and static fs applied.
> 
> Output:
> 
> ```
> <!DOCTYPE html>
> <html>
> 
> <head>
>     <title>fresh-project</title>
>     <meta charset="utf-8" />
>     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
>     <link rel="stylesheet" href="/styles.css" />
>     <link rel="modulepreload" href="/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/deserializer.js" />
>     <link rel="modulepreload" href="/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/signals.js" />
>     <link rel="modulepreload" href="/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/main.js" />
>     <link rel="modulepreload" href="/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/island-counter.js" />
> </head>
> 
> <body>
>     <div class="px-4 py-8 mx-auto bg-[#86efac]">
>         <div class="max-w-screen-md mx-auto flex flex-col items-center justify-center">the Fresh logo: a sliced lemon dripping with juice
>             <h1 class="text-4xl font-bold">Welcome to Fresh</h1>
>             <p class="my-4">Try updating this message in the<code class="mx-2">./routes/index.tsx</code> file, and
>                 refresh.</p>
>             
>             <div class="flex gap-8 py-6">
>                 <button disabled class="px-2 py-1 border-gray-500 border-2 rounded bg-white hover:bg-gray-200 transition-colors">-1</button>
>                 <p class="text-3xl tabular-nums">3</p>
>                 <button disabled class="px-2 py-1 border-gray-500 border-2 rounded bg-white hover:bg-gray-200 transition-colors">+1</button>
>             </div>
>             
>         </div>
>     </div>
>     <script id="__FRSH_STATE_4bac2f0c-8d08-4e0a-885a-c51491ebafcc" type="application/json"
>         nonce="341dac263ecd40cd91b676254f2f435c">
>         {"v":[[{"count":{"_f":"s","v":3}}],[]]}
>     </script>
>     <script type="module" nonce="341dac263ecd40cd91b676254f2f435c">
>         import { deserialize } from "/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/deserializer.js";import { signal } from "/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/signals.js";const ST = document.getElementById("__FRSH_STATE_4bac2f0c-8d08-4e0a-885a-c51491ebafcc").textContent;const STATE = deserialize(ST, signal);import { revive } from "/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/main.js";import counter_default from "/_frsh/js/d9d51bae574d9c3d4563416258222b74e69df4ee/island-counter.js";const propsArr = typeof STATE !== "undefined" ? STATE[0] : [];revive({counter_default:counter_default,}, propsArr);
>     </script>
> </body>
> 
> </html>
> ```
> 
> Note: This issue also affects a decorator PR I recently submitted.

Amazing! Will test soon.

Just to clarify one thing, do I need to set the file system to be the Deno's Compile one? or there are a command that make it work out of the box?

**Comment by nyannyacha** [comment]
@mcandeia 

> Just to clarify one thing, do I need to set the file system to be the Deno's Compile one? or there are a command that make it work out of the box?

Oh wow, sorry, I'm not a native, so I guess that sentence was ambiguous. 😂
I've confirmed that applying the patch I posted above will render HTML content on your fork.

(However, I've only checked with Postman, so I can't guarantee that everything works for the fresh example)

**Comment by mcandeia** [comment]
> @mcandeia
> 
> > Just to clarify one thing, do I need to set the file system to be the Deno's Compile one? or there are a command that make it work out of the box?
> 
> Oh wow, sorry, I'm not a native, so I guess that sentence was ambiguous. 😂 I've confirmed that applying the patch I posted above will render HTML content on your fork.
> 
> (However, I've only checked with Postman, so I can't guarantee that everything works for the fresh example)

Oh, I mean, I had to change the following line: https://github.com/deco-cx/edge-runtime/pull/6/files#diff-cce0097be8934b699189157e13e2c8eed2421b37691ab0d0cb25640dcc0f741bL369

is it still necessary or there is a command to link the static file system (like --static *)? To be honest, it would be amazing to have a feature to pass the filesystem inline as a hashmap

{"FILE_PATH": "FILE_CONTENT_BYTES" }

**Comment by mcandeia** [comment]
image

Yeah, its happening!

**Comment by nyannyacha** [comment]
@mcandeia 

> Oh, I mean, I had to change the following line: https://github.com/deco-cx/edge-runtime/pull/6/files#diff-cce0097be8934b699189157e13e2c8eed2421b37691ab0d0cb25640dcc0f741bL369

static fs is a relaxation of some of the sandboxing measures to allow users of Supabase's hosted environment to access only proven safe paths through the file system API.

This does not appear to be currently compatible with packages such as the `fresh` package, as the mounting points are fixed, as you seen above 😓

> is it still necessary or there is a command to link the static file system (like --static *)? To be honest, it would be amazing to have a feature to pass the filesystem inline as a hashmap

Instead of specifying a series of `--static` flags, it might also be a good idea to specify a simple plain text file that lists what to allow within the runtime 😋

Thanks for the feedback!

**Comment by mcandeia** [comment]
@nyannyacha Want me to open a PR with the modifications you mentioned plus this one? I will be glad to have my first contribution

**Comment by nyannyacha** [comment]
@mcandeia 

Yes I would be happy for you to do that 😁

But in the case of static fs modification PR, it was written by @andreespirela, so I think it would be better to hear his thoughts first.

**Comment by andreespirela** [comment]
@mcandeia We're discussing the how-to internally. It doesn't make sense to disable StaticFS, and changing `import.meta.url` is not the way to go. More to come... I'm trying to make sure fresh runs smooth.

**Comment by mcandeia** [comment]
> @mcandeia We're discussing the how-to internally. It doesn't make sense to disable StaticFS, and changing `import.meta.url` is not the way to go. More to come... I'm trying to make sure fresh runs smooth.

If I can share my thoughts: I would love to have a way to pass the filesystem inline as a hashmap. Or even as a "lazy" structure. So we can have two separate things:

- Eszip having the .ts/.tsx files
- A lazy filesystem which would use NFS/Block Storage.

My 50 cents

**Comment by mcandeia** [comment]
Hello guys @andreespirela @nyannyacha, any news?

**Comment by mcandeia** [comment]
Guys @laktek @andreespirela and @nyannyacha, I think the scope of this PR now has changed completely, initially it should be only for supporting JSX import source and I think it was done, just need to fix the missing emitter options as @nyannyacha did.

Ins't better to just create a separate issue for tracking filesystem-related things?

Running fresh projects is out of scope of this PR, it was just an example, supporting JSX is a good start for that

**Comment by nyannyacha** [comment]
> Guys @laktek @andreespirela and @nyannyacha, I think the scope of this PR now has changed completely, initially it should be only for supporting JSX import source and I think it was done, just need to fix the missing emitter options as @nyannyacha said.
> 
> Ins't better to just create a separate issue for tracking filesystem-related things?
> 
> Running fresh projects is out of scope of this PR, it was just an example, supporting JSX is a good start for that
> 
> 

Yeah, now that I think about it, this was PR for JSX support.

I think I've gotten away from the point of this PR, as I thought the fresh package case was the main use case.

So I think you're right, we need to separate the static fs issue.

**Review by nyannyacha — APPROVED** [review]

**Pull request merged**